server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceRegistry.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.search.aggregations.support;
+
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.IndexGeoPointFieldData;
+import org.elasticsearch.index.fielddata.IndexNumericFieldData;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.RangeFieldMapper;
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+/*
+This is a _very_ crude prototype for the ValuesSourceRegistry which basically hard-codes everything.  The intent is to define the API
+for aggregations using the registry to resolve aggregators.
+ */
+public enum ValuesSourceRegistry {
+    INSTANCE {
```

Comment:
Sorry about the funky mutable enum singleton pattern, I just needed something that wasn't going to get in the way for prototyping.  Happy to use a different implementation before going to production. 

This is intended to illustrate the core API for the new registry pattern.  We'll probably need to add one or two more registered components, for the script and unmapped branches in `ValuesSourceConfig.resolve()`, but in general this is how it should work: for each aggregation family, a mapping of predicates to `ValuesSourceType`s and a mapping of `ValuesSourceType`s to `AggregatorSuppliers` (where the actual aggregator supplier implements an aggregation family specific sub-interface).  

This is the heart of the plan, so please raise any concerns with this, no matter how trivial.